hedera-node/hedera-app/src/main/java/com/hedera/node/app/tss/TssBaseServiceImpl.java

```diff
@@ -380,12 +391,255 @@ public TssMessage getTssMessageFromBytes(Bytes wrap, TssParticipantDirectory dir
+                                                        final var tssMessage = TssMessageTransactionBody.newBuilder()
+                                                                .sourceRosterHash(activeRosterHash)
+                                                                .targetRosterHash(candidateRosterHash)
+                                                                .shareIndex(shareIndex.getAndAdd(1))
+                                                                .tssMessage(Bytes.wrap(msg.toBytes()))
+                                                                .build();
+                                                        // need to use consensusNow here
+                                                        tssSubmissions.submitTssMessage(tssMessage, consensusNow);
+                                                        haveSentMessageForTargetRoster = true;
+                                                    },
+                                                    tssLibraryExecutor)
+                                            .exceptionally(e -> {
+                                                log.error("Error generating tssMessage", e);
+                                                return null;
+                                            });
+                                }
+                            }
+                        }
+                    })
+                    .exceptionally(e -> {
+                        log.error("Error validating threshold TSS messages", e);
+                        return null;
+                    });
+        }
+
+        private void reset() {
+            newRosterToKey = RosterToKey.NONE;
+            haveSentMessageForTargetRoster = false;
+            haveSentVoteForTargetRoster = false;
+        }
+
+        private CompletableFuture<Boolean> validateThresholdTssMessages() {
+            final var participantDirectory = tssDirectoryAccessor.activeParticipantDirectory();
+            final var tssMessageBodies = info.tssMessages();
+            return CompletableFuture.supplyAsync(
+                    () -> {
+                        final var tssMessages = validateTssMessages(tssMessageBodies, participantDirectory, tssLibrary);
+                        return isThresholdMet(tssMessages, participantDirectory);
+                    },
+                    tssLibraryExecutor);
+        }
+
+        private void validateThresholdEncryptionKeysReached(@NonNull Bytes rosterHash) {
+            var numTssEncryptionKeys = info.encryptionKeys().size();
+            final var thresholdReached = numTssEncryptionKeys
+                    >= (2 * requireNonNull(info.activeRoster()).rosterEntries().size()) / 3;
+            if (thresholdReached) {
+                newKeyingStatus = TssKeyingStatus.WAITING_FOR_THRESHOLD_TSS_MESSAGES;
+            } else {
+                // TODO: Create the encryption key for self if not present
```

Comment:
When we address this TODO we may want to consider that we should submit our TSS public key even if the threshold is already reached. If we do, we may need to put the checks for submitting the TSS public key into `computeNewStatus` or `manageTssStatus` to be checked every time until our public key is in state.